### PR TITLE
[BPK-3567]: Upgrade RN to 0.61.5

### PR DESCRIPTION
**Pull request by olliecurtis**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-react-native/blob/master/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `UNRELEASED.md`
+ [x] `README.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: packages/react-native-bpk-theming/src/BpkThemeProvider.js [hunk 1]
```diff
@@ -22,6 +22,16 @@ import createWithTheme from './create-with-theme';
 
 const BpkThemeProvider = ThemeProvider;
 
+// TODO: Upgrade theming package and remove this: https://gojira.skyscanner.net/browse/BPK-3639
```
Comment (tiagoengel) [inline packages/react-native-bpk-theming/src/BpkThemeProvider.js]: I added this because this causes a lot of warnings making it really hard to see anything important. The same thing happens with `@skyscanner/react-native-transitiongroup` and `recyclerlistview` but those are only used in the carousel and not really a big issue. I have also added cards to upgrade those

Comment (olliecurtis) [reply packages/react-native-bpk-theming/src/BpkThemeProvider.js]: Should we be having a link to our internal tools here in the source?

File: .flowconfig [hunk 2]
```diff
@@ -23,6 +23,9 @@ node_modules/warning/.*
 ; Flow doesn't support platforms
 .*/Libraries/Utilities/HMRLoadingView.js
 
+; Flow doesn't support platforms
+.*/Libraries/Utilities/LoadingView.js
```
Comment (olliecurtis) [inline .flowconfig]: Ah that's how you get around react itself throwing errors!

File: packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js [hunk 3]
```diff
@@ -101,7 +101,8 @@ class AnimateAndFade extends Component<Props> {
     this.measure(captureHeightAndAnimate);
   }
 
-  componentWillReceiveProps(nextProps: Props) {
+  // eslint-disable-next-line camelcase
+  UNSAFE_componentWillReceiveProps(nextProps: Props) {
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js]: Should we be fixing this now rather than waiting for the point it is removed?

File: packages/react-native-bpk-component-banner-alert/src/__snapshots__/BpkBannerAlert-test.android.js.snap [hunk 4]
```diff
@@ -317,7 +317,7 @@ exports[`Android BpkBannerAlert dark mode should render correctly with dismissab
             accessibilityRole="button"
             accessibilityStates={Array []}
             accessible={true}
-            clickable={true}
+            focusable={true}
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-banner-alert/src/__snapshots__/BpkBannerAlert-test.android.js.snap]: Is this change intentional?

File: packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap [hunk 5]
```diff
@@ -19,7 +19,7 @@ exports[`Android API levels Android API<21 should render without ripple 1`] = `
     accessibilityRole="button"
     accessibilityStates={Array []}
     accessible={true}
-    clickable={true}
+    focusable={true}
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap]: Same as previous comment

File: packages/react-native-bpk-component-card/src/__snapshots__/BpkCard-test.android.js.snap [hunk 6]
```diff
@@ -16,7 +16,7 @@ exports[`Android BpkCard dark mode should render correctly 1`] = `
     accessibilityLabel="Example Card"
     accessibilityRole="button"
     accessible={true}
-    clickable={true}
+    focusable={true}
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-card/src/__snapshots__/BpkCard-test.android.js.snap]: Same as previous comment

File: packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js [hunk 7]
```diff
@@ -104,7 +104,8 @@ class BpkHorizontalNav extends React.Component<
     };
   }
 
-  componentWillReceiveProps(nextProps) {
+  // eslint-disable-next-line camelcase
+  UNSAFE_componentWillReceiveProps(nextProps) {
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js]: Should we fix this now to use the correct methods?

File: packages/react-native-bpk-component-navigation-bar/src/isIphoneX.js [hunk 8]
```diff
@@ -23,8 +23,12 @@ export default (() => {
   if (Platform.OS === 'android') {
     return false;
   }
+  const windowDimensions = DeviceInfo.getConstants().Dimensions.window;
+  if (!windowDimensions) {
```
Comment (olliecurtis) [inline packages/react-native-bpk-component-navigation-bar/src/isIphoneX.js]: Could you clarify the logic and what is happening here?

File: packages/react-native-bpk-theming/src/BpkThemeProvider.js [hunk 9]
```diff
@@ -22,6 +22,16 @@ import createWithTheme from './create-with-theme';
 
 const BpkThemeProvider = ThemeProvider;
 
+// TODO (BPK-3639): Upgrade theming package and remove this
+const cwm = ThemeProvider.prototype.componentWillMount;
```
Comment (olliecurtis) [inline packages/react-native-bpk-theming/src/BpkThemeProvider.js]: Could you clarify what this logic is doing here?

File: .flowconfig (hunk 2 above)
Comment (tiagoengel) [reply .flowconfig]: Yeah, in this case is flow's inability to understand platform imports, `.ios` and `.android`, so we need to ignore this file.

File: packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js (hunk 3 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js]: We should fix deprecations as soon as possible, this will be an error in the next release so is better to update now while we still have time to catch any problems. Plus this causes a very big an annoying warning.

File: packages/react-native-bpk-component-banner-alert/src/__snapshots__/BpkBannerAlert-test.android.js.snap (hunk 4 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-banner-alert/src/__snapshots__/BpkBannerAlert-test.android.js.snap]: Yes, this is a change in how RN's `TouchableNativeFeedback` is rendered. It doesn't affect the actual usage of the component.

File: packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap (hunk 5 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap]: yes

File: packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js (hunk 3 above)
Comment (olliecurtis) [reply packages/react-native-bpk-component-banner-alert/src/AnimateAndFade.js]: Can we/should we be doing this in this PR? Or do we have an action to do that later?

File: packages/react-native-bpk-component-navigation-bar/src/isIphoneX.js (hunk 8 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-navigation-bar/src/isIphoneX.js]: The interface for this has changed (https://github.com/facebook/react-native/pull/24878/files) and `window` is an `iOS` only property, and it can be null here because this file is not platform specific. So we first need to check if it is present before we use it. The end result should be the same as before

Comment (olliecurtis) [reply packages/react-native-bpk-component-navigation-bar/src/isIphoneX.js]: Ok sounds great 😄

File: packages/react-native-bpk-theming/src/BpkThemeProvider.js (hunk 9 above)
Comment (tiagoengel) [reply packages/react-native-bpk-theming/src/BpkThemeProvider.js]: I'm just renaming `componentWillMount` and `componentWillReceiveProps` to `UNSAFE_componentWillMount` and `UNSAFE_componentWillReceiveProps` to remove the deprecation warnings. Once we update the theming package we can remove this, but because we don't expose it directly to consumers we can do this little hack to remove the warning for now and avoid polluting the app with even more warnings.

Comment (olliecurtis) [reply packages/react-native-bpk-theming/src/BpkThemeProvider.js]: I am wondering if we should be hacking this instead of just moving to the correct replacements and fixes for this?

File: packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js (hunk 7 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js]: We should fix this as soon as we can, but I don't think we should as part of this update because is more changes to introduce errors in an already big change. So I'd rather do the RN update first and then look at removing these methods.

Comment (olliecurtis) [reply packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js]: Ok sounds good - we should make an issue for this 😄

File: packages/react-native-bpk-theming/src/BpkThemeProvider.js (hunk 9 above)
Comment (tiagoengel) [reply packages/react-native-bpk-theming/src/BpkThemeProvider.js]: The only thing we can do to remove this is upgrading the theming lib, and it's a breaking change, so I'd rather not do that with the RN upgrade.

File: packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js (hunk 7 above)
Comment (tiagoengel) [reply packages/react-native-bpk-component-horizontal-nav/src/BpkHorizontalNav.js]: We have one in the backlog already to remove deprecated methods :)
